### PR TITLE
ddl/ingest: move backend ctx out of checkpoint manager

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref #42164

Problem Summary:

### What changed and how does it work?

- Move backend ctx out of checkpoint manager and remove `FlushCtrl`.
- Remove incomplete and unstable test `TestAddIndexIngestRecoverPartition`.
- Add a failure test during ingest.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: tests/realtikvtest/addindextest4/failure_test.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		store := realtikvtest.CreateMockStoreAndSetup(t)
+		realtikvtest.RetainOldData = false
+		tk := testkit.NewTestKit(t, store)
+		tk.MustExec("use addindexlit;")
+		<-block
+		tk.MustExec("admin check table t;")
+	})
+}
+
+func ExecuteBlocks(t *testing.T, fns ...func()) {
```
Comment (lance6716) [inline tests/realtikvtest/addindextest4/failure_test.go:105]: Can you explain the purpose of this function? Seems it runs `fns` in different processes. But `fns` are not executed until the previous one exited, can we run them one-by-one in one process?

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: tests/realtikvtest/addindextest4/failure_test.go (hunk 1 above)
Comment (tangenta) [reply tests/realtikvtest/addindextest4/failure_test.go:105]: > can we run them one-by-one in one process?

No. I want to mock the case like `kill -9` with `os.Exit(0)`, and see if the checkpoint manager can handle this correctly.

**Review by lance6716 — APPROVED** [review]

File: pkg/ddl/ingest/checkpoint.go [hunk 2]
```diff
@@ -215,15 +212,17 @@ func (s *CheckpointManager) UpdateTotalKeys(taskID int, delta int, last bool) {
...
 	s.mu.Lock()
 	cp := s.checkpoints[taskID]
 	cp.writtenKeys += delta
 	s.mu.Unlock()
+}
 
-	flushed, imported, err := s.flushCtrl.Flush(FlushModeAuto)
-	if !flushed || err != nil {
-		return err
+// AdvanceWatermark advances the watermark according flushed or imported status.
```
Comment (tangenta) [inline pkg/ddl/ingest/checkpoint.go:222]:
```suggestion
// AdvanceWatermark advances the watermark according to flushed or imported status.
```

**Review by D3Hunter — APPROVED** [review]

**Comment by tangenta** [comment]
Blocked by #54543

**Pull request merged**